shared/lib-hoppy/hoppy/async_hoppy_client.py

```diff
@@ -0,0 +1,204 @@
+import asyncio
+import json
+import logging
+import time
+import uuid
+
+from hoppy.async_consumer import AsyncConsumer
+from hoppy.async_publisher import AsyncPublisher
+from hoppy.exception import ResponseException
+from hoppy.hoppy_properties import ExchangeProperties, QueueProperties
+from pika import BasicProperties
+
+MAX_RETRIES_REACHED = "Max retries reached"
+TIMED_OUT = "Request timed out"
+UNDECODABLE = "JSON response could not be parsed"
+UNEXPECTED_PUBLISH_ERROR = "Unexpected Error Caught While Publishing"
+
+
+class AsyncHoppyClient:
+    responses = {}
+    rejected = {}
+
+    def __init__(self,
+                 name: str,
+                 app_id: str,
+                 config: dict,
+                 exchange_properties: ExchangeProperties,
+                 request_queue_properties: QueueProperties,
+                 reply_queue_properties: QueueProperties,
+                 request_routing_key: str,
+                 reply_routing_key: str,
+                 max_latency: int = 30,
+                 response_reject_and_requeue_attempts: int = 3):
+        self.name = name
+        self.app_id = app_id
+        self.config = config
+
+        self.exchange_properties = exchange_properties
+        self.request_queue_properties = request_queue_properties
+        self.reply_queue_properties = reply_queue_properties
+
+        self.max_latency = max_latency
+        self.response_reject_and_requeue_attempts = response_reject_and_requeue_attempts
+
+        self.async_publisher = AsyncPublisher(config=self.config,
+                                              exchange_properties=exchange_properties,
+                                              queue_properties=request_queue_properties,
+                                              routing_key=request_routing_key)
+
+        self.async_consumer = AsyncConsumer(config=self.config,
+                                            exchange_properties=exchange_properties,
+                                            queue_properties=reply_queue_properties,
+                                            routing_key=reply_routing_key,
+                                            reply_callback=self._on_reply)
+
+    def start(self, loop):
+        self.async_publisher.connect(loop)
+        self.async_consumer.connect(loop)
+
+    def stop(self):
+        self.async_publisher.stop()
+        self.async_consumer.stop()
+
+    async def make_request(self, request_id, body):
+        correlation_id = str(uuid.uuid4())
+        logging.info(
+            f"event=requestStarted id={request_id} "
+            f"client={self.name} "
+            f"queue={self.request_queue_properties.name} "
+            f"correlation_id={correlation_id}")
+        self.responses[correlation_id] = None
```

Comment:
It took me a while to realize that this dictionary is actually just `AsyncHoppyClient.responses`. Is there a reason to make it a class-level dictionary, rather than initiate an instance-level in `__init__`?

And my second question is, what is the benefit of inserting a None at the beginning of the request here, only to replace it with the response (or a JSON error) later?

Reply:
Q1: No reason! Makes sense to make it instance level. 
Q2: I need to insert a value of None for the correlation_id in the responses at the beginning of the request because the client needs to keep track of the correlation_id's for which it is responsible. If we had two clients both listening to the same queue for replies, we don't want it processing correlation_ids for requests it didnt send out.